### DISCUSSION TITLE
Create a "stack" of images centered with border

### DISCUSSION THREAD

**Discussion by GeneaCoder** [ImageMagick/ImageMagick#2293] [Help] [answered]
Version: ImageMagick 6.9.10-68 Q16 x86_64 2020-04-13 https://imagemagick.org

I'm trying to combine 4 images together with some borders/spacers.

```
######################   #-black border and center spacer
##<----Image #1---->##
##<--White Spacer-->##   group 1 = img1, white spacer, img2
##<----Image #2---->##
######################
####<--Image #3-->####
####<White Spacer>####   group 2 = img3, white spacer, img3
####<--Image #4-->####
######################
```

Some specs...
Border is black, uniform (currently 20px)
Black spacer in middle (currently 20px high)
White Spacers have 10px height, variable width
Images 1 & 2 are ALWAYS the same width, variable height (Spacer 1 has same width)
Images 3 & 4 are ALWAYS the same width, variable height (Spacer 2 has same width)
Images 1 & 2 are RARELY the same width as 3 & 4

I want everything centered L to R.

Attempt 1 - I tried creating the white and black spacers and then `+append` all 7 together on a black background. I failed (almost assuredly because I'm doing it wrong). I never got to the "add a border" step of the process.
Attempt 2 - I calculated everything and used `composite -geometry`. This works...slowly.

```
convert -size 2168x10 xc:white whsp2168.png
convert -size 2137x10 xc:white whsp2137.png
convert -size 2208x1159 xc:black stacked.png
composite -geometry +20+20  image1.png stacked.png stacked.png
composite -geometry +20+351 whsp2168.png stacked.png stacked.png
composite -geometry +20+361 image2.png stacked.png stacked.png
composite -geometry +36+617 image3.png stacked.png stacked.png
composite -geometry +36+948 whsp2137.png stacked.png stacked.png
composite -geometry +36+958 image4.png stacked.png stacked.png
```

In the above example images 1 & 2 are 2168px wide, 3 & 4 are 2137px wide. (2137 being an odd number, the "center" is off by 1/2 pixel...I'm fine with this.)

What I'm looking for is as few commands as possible. This would (hopefully) include NOT having to create white spacers every time. I'd like to create a 1x1 white dot and distort it to size (2168x10 for example) during the process. However, first priority is putting them all together in one shot.

I'm running this on a webserver via php. I have (or can get) all height and width info needed. Images 1-4 are variable, but generally 1500-2500 pixels wide and 150-250 pixels high. When all is said and done, I'll also need to convert this image back to jpg form.

Thank you for reading! It is appreciated!

**Comment by snibgo** [comment, marked as answer]
How about this. Windows BAT syntax. Adjust it for PHP or whatever.
```
convert ^
  -gravity Center ^
  ( ^
    toes.png ^
    -size 1x10 xc:White ^
    toes.png ^
    -background White ^
    -append ^
  ) ^
  -size 1x20 xc:Black ^
  ( ^
    rose: ^
    -size 1x10 xc:White ^
    rose: ^
    -background White ^
    -append ^
  ) ^
  -background Black ^
  -append ^
  -bordercolor Black -border 20 ^
  stack2.jpg
```
![stack2](https://user-images.githubusercontent.com/33812368/87856876-f29d9380-c919-11ea-94ed-531c8cf62a8f.jpg)

**Reply by GeneaCoder** [reply]
Sometimes I wonder, "Why is that frisbee getting bigger? And then it hits me!"

This makes sense. I can see now why my attempt 1 with `-append` failed. But I have to say, when I put it into my code, it also failed. It didn't take long to realize the (parenthesis) were the cause of my problems. Escaping them did not work. Escaping them _correctly_ finally did work.

For anyone who may find this on a search...
Escaping parenthesis () in ImageMagick via php exec command needs to be aware of the following... (from https://imagemagick.org/script/command-line-processing.php)

> Each parenthesis (or escaped parenthesis) must have spaces on either side.

`convert -gravity Center \(image1.png blah blah blah -append\) blah blah`
...will not work, however...
`convert -gravity Center \( image1.png blah blah blah -append \) blah blah`
...works just fine.

Thank you, snibgo! Thank you, thank you, thank you!

P.S. `-size 1x10 xc:White` is a SEXY part of this solution!

**Comment by snibgo** [comment]
Yes, parentheses in IM need a space on both sides.

Yes, the one-pixel-wide spacers are key. "-append" pads the width of these using the "-background" colour.
